api/base/exceptions.py

```diff
@@ -26,28 +46,33 @@ def json_api_exception_handler(exc, context):
                 }
             ])
         elif isinstance(message, dict):
-            for error_key, error_description in message.iteritems():
-                if error_key in top_level_error_keys:
-                    errors.append({error_key: error_description})
-                elif error_key in resource_object_identifiers:
-                    if isinstance(error_description, basestring):
-                        error_description = [error_description]
-                    errors.extend([{'source': {'pointer': '/data/' + error_key}, 'detail': reason} for reason in error_description])
-                elif error_key == 'attributes':
-                    if isinstance(error_description, list):
-                        errors.extend([{'source': {'pointer': '/data/' + error_key}, 'detail': reason} for reason in error_description])
-                else:
-                    if isinstance(error_description, basestring):
-                        error_description = [error_description]
-                    errors.extend([{'source': {'pointer': '/data/attributes/' + error_key}, 'detail': reason} for reason in error_description])
-
+            dict_error_formatting(errors, message)
         else:
             if isinstance(message, basestring):
                 message = [message]
-            errors.extend([{'detail': error} for error in message])
+            for error in message:
+                if isinstance(error, dict):
+                    dict_error_formatting(errors, error)
+                else:
+                    errors.append({'detail': error})
 
         response.data = {'errors': errors}
 
+        # For bulk operations: If 400 error, return request data with response.
```

Comment:
Is it possible to match on the type of `exc` rather than doing string comparison for `'non_field_errors'`?